plugins/inputs/win_wmi/win_wmi.go

```diff
@@ -0,0 +1,268 @@
+//go:build windows
+// +build windows
+
+package win_wmi
+
+import (
+	_ "embed"
+	"fmt"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	ole "github.com/go-ole/go-ole"
+	"github.com/go-ole/go-ole/oleutil"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/filter"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+// Query struct
+type Query struct {
+	Query                string
+	Namespace            string   `toml:"Namespace"`
+	ClassName            string   `toml:"ClassName"`
+	Properties           []string `toml:"Properties"`
+	Filter               string   `toml:"Filter"`
+	TagPropertiesInclude []string `toml:"TagPropertiesInclude"`
+	tagFilter            filter.Filter
+}
+
+// Wmi struct
+type Wmi struct {
+	Queries []Query `toml:"query"`
+	Log     telegraf.Logger
+}
+
+// S_FALSE is returned by CoInitializeEx if it was already called on this thread.
+const sFalse = 0x00000001
+
+func oleInt64(item *ole.IDispatch, prop string) (int64, error) {
+	v, err := oleutil.GetProperty(item, prop)
+	if err != nil {
+		return 0, err
+	}
+	defer func() { _ = v.Clear() }()
+
+	i := v.Val
+	return i, nil
+}
+
+// Init function
+func (s *Wmi) Init() error {
+	if err := CompileInputs(s); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Description function
+func (s *Wmi) Description() string {
+	return "returns WMI query results as metrics"
+}
+
+// SampleConfig function
+func (s *Wmi) SampleConfig() string {
+	return sampleConfig
+}
+
+func CompileInputs(s *Wmi) error {
+	BuildWqlStatements(s)
+	if err := CompileTagFilters(s); err != nil {
+		return err
+	}
+	return nil
+}
+
+func CompileTagFilters(s *Wmi) error {
+	var err error
+	for i, q := range s.Queries {
+		s.Queries[i].tagFilter, err = CompileTagFilter(q)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func CompileTagFilter(q Query) (filter.Filter, error) {
+	tagFilter, err := filter.NewIncludeExcludeFilterDefaults(q.TagPropertiesInclude, nil, false, false)
+	if err != nil {
+		return nil, fmt.Errorf("creating tag filter failed: %v", err)
+	}
+	return tagFilter, nil
+}
+
+func BuildWqlStatements(s *Wmi) {
+	for i, q := range s.Queries {
+		s.Queries[i].Query = BuildWqlStatement(q)
+	}
+}
+
+// build a WMI query from input configuration
+func BuildWqlStatement(q Query) string {
+	wql := fmt.Sprintf("SELECT %s FROM %s", strings.Join(q.Properties, ", "), q.ClassName)
+	if len(q.Filter) > 0 {
+		wql = fmt.Sprintf("%s WHERE %s", wql, q.Filter)
+	}
+	return wql
+}
+
+func DoQuery(q Query, acc telegraf.Accumulator) error {
```

Comment:
What do you think about moving this to a class method?
```suggestion
func (q *Query) DoQuery(acc telegraf.Accumulator) error {
```

Reply:
Thanks! I did this.